File: max_in_sorted_array.js
#!/usr/bin/env node

var _ = require('underscore');

var range = _.range(1, 18); // 1..17
var data = _.rest(range, 3).concat(_.first(range,3));

var find_max = function(a) {
  var last = a[a.length - 1];
  if (a[0] <= last) {
    return last;
  }
  var pivot = parseInt(a.length / 2);
  if (a[pivot] > last) {
    return find_max(a.slice(pivot, a.length));
  } else {
    return find_max(a.slice(0, pivot));
  }
}

console.log("input:", data.join(','));
console.log("max:", find_max(data));

/*

input: 4,5,6,7,8,9,10,11,12,13,14,15,16,17,1,2,3
max: 17

*/

File: max_in_sorted_array.rb
#!/usr/bin/env ruby

data = (1..17).to_a.rotate!(3)

def find_max(a)
  return a.last if a.first <= a.last
  pivot = a.size / 2
  if a[pivot] > a.last
    return find_max(a[pivot..-1])
  else
    return find_max(a[0..(pivot - 1)])
  end
end

puts "input: #{data.join(',')}"
puts "max: #{find_max(data)}"

__END__

input: 4,5,6,7,8,9,10,11,12,13,14,15,16,17,1,2,3
max: 17
